docs/blueprints/ge_refactoring.md

```diff
@@ -0,0 +1,67 @@
+Hello!
+
+This article is all about our vision how to make graceful exit code a bit readable and testable.
+
+So, main logic is placed in Worker (storj/storagenode/gracefulexit/worker.go)
+
+This module has giant list of dependencies, goroutines, large methods, lack of comments and tests.
+
+Moreover, in this place we combine all this logic - database calls, business data processing, different transport level dialings, infinite loops, goroutines.
+
+Size of this code is too large to understand it easily. Sometimes we faced such things like missing returns and wrong error handling.
+
+All this factors won't allow us to be sure that it works fine even if there won't be any bug tickets for some time.
+
+We all understand that small pieces of code are much more readable and its easier to test it and to find potential bug just by reading it.
+
+So, our vision is next:
+
+1. **Worker** - its main responsibility to run whole process. It should run goroutines, listen messages, make all needed dialing.
+It should not depend on some concrete implementation, have references to database, etc.
+This worker should have dependencies interfaces, to be able to mock them in future.
+2. **Service** - service should gather data from database, process them or save them in database.
+
+I would expect Service layer to be an interface also, to be able to mock it
+
+```go
+type GracefulExit interface {
+	TransferPiece(ctx context.Context, satelliteID storj.NodeID, transferPiece *pb.TransferPiece, c gracefulExitStream) error
+	DeleteOnePiece(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) error
+	DeletePiece(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) error
+	DeleteAllPieces(ctx context.Context, satelliteID storj.NodeID) error
+	Fail(ctx context.Context, satelliteID storj.NodeID, reason pb.ExitFailed_Reason, exitFailedBytes []byte) error
+	Complete(ctx context.Context, satelliteID storj.NodeID, exitFailedBytes []byte, wait func()) error
+	CancelGracefulExit(ctx context.Context, satelliteID storj.NodeID) error
+}
+```
+its implementation could have all db related depenncies 
+```go
+var _ GracefulExit = (*Service)(nil)
+type Service struct {
+	log *zap.Logger
+
+	store       *pieces.Store
+	trust       *trust.Pool
+	ecclient    ecclient.Client
```

Comment:
To me the ec client should be part of the worker, not the service.

Reply:
I don't know if I entirely agree. I don't think it should be a part of the worker, but I also understand not wanting it to be a crucial part of the service that updates the DB since that might get confusing. Maybe we can consider having another service that has the responsibility of transferring the piece (e.g. GracefulExitTransferService) and keep the rest of the logic in this service (e.g. GracefulExitStateService)

Reply:
i absolutely agree that ecclient.Client should not be in the service, unfortunately, i didn't have an opportunity to move it out of service, because of lack of time. In that case i would have to broke TransferPiece method on 2 different methods or pass it as callback, etc.

Reply:
If it is something that we want to do in the future (moving ecclient.Client out of the service), we should make sure to mention it somewhere in the design doc, even if we do not implement it in the initial refactor - that way there is a record for future work